third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp

```diff
@@ -611,28 +611,71 @@ LogicalResult PointerCanonicalizer::rewriteForOp(scf::ForOp forOp,
                                                  Location curLoc,
                                                  OpOperand *curOperand,
                                                  Value &nextPtr) {
-  size_t operandNum = curOperand->getOperandNumber();
   FatPtr fatPtr = pointers[curOperand->get()];
   Value offset = fatPtr.offset;
   Value basePtr = fatPtr.basePtr;
 
-  // Replace the forOp with two additional argument (i.e., the curOperand's
-  // scalar pointer and the offset)
-  Value tensorPtr = createTensorPointer(fatPtr, curLoc);
-  auto newForOp =
-      replaceForOpWithNewSignature(rewriter, forOp, {basePtr, offset});
+  size_t operandNum = curOperand->getOperandNumber();
+  auto numLeadingInitsIncludingCurOp =
+      operandNum + 1 - forOp.getNumControlOperands();
+
+  SmallVector<Value> newInitArgs(
+      forOp.getInitArgs().take_front(numLeadingInitsIncludingCurOp));
+  newInitArgs.append({basePtr, offset});
+  auto trailingInits =
+      forOp.getInitArgs().drop_front(numLeadingInitsIncludingCurOp);
+  newInitArgs.append(trailingInits.begin(), trailingInits.end());
+
+  scf::ForOp newForOp;
+  {
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPoint(forOp);
+    newForOp = rewriter.create<scf::ForOp>(
+        forOp.getLoc(), forOp.getLowerBound(), forOp.getUpperBound(),
+        forOp.getStep(), newInitArgs);
+    newForOp->setAttrs(forOp->getAttrs());
+    newForOp.getBody()->erase();
+    newForOp.getRegion().getBlocks().splice(
+        newForOp.getRegion().getBlocks().begin(),
+        forOp.getRegion().getBlocks());
+
+    newForOp.getBody()->insertArgument(numLeadingInitsIncludingCurOp + 1,
+                                       offset.getType(), offset.getLoc());
+    newForOp.getBody()->insertArgument(numLeadingInitsIncludingCurOp + 1,
+                                       basePtr.getType(), basePtr.getLoc());
+
+    for (auto [src, tgt] : llvm::zip(
+             forOp.getResults().take_front(numLeadingInitsIncludingCurOp),
+             newForOp.getResults().take_front(numLeadingInitsIncludingCurOp)))
+      rewriter.replaceAllUsesWith(src, tgt);
+    for (auto [src, tgt] :
+         llvm::zip(forOp.getResults().drop_front(numLeadingInitsIncludingCurOp),
+                   newForOp.getResults().drop_front(
+                       numLeadingInitsIncludingCurOp + 2)))
+      rewriter.replaceAllUsesWith(src, tgt);
+
+    newForOp.getBody()->getTerminator()->insertOperands(
+        numLeadingInitsIncludingCurOp,
+        newForOp.getRegionIterArg(numLeadingInitsIncludingCurOp));
+    newForOp.getBody()->getTerminator()->insertOperands(
+        numLeadingInitsIncludingCurOp + 1,
+        newForOp.getRegionIterArg(numLeadingInitsIncludingCurOp + 1));
+  }
+
   rewriteOpMap[forOp] = newForOp;
 
-  newForOp->setOperand(operandNum, tensorPtr);
+  {
```

Comment:
Basically, if you remove the current operand from the the loop arguments, you won't need this anymore. This will be removed by canonicalization anyway, but it's here only because we have to pass "something" as the current operand.